### PR TITLE
Update connection management documentation

### LINKED ISSUE #1675: Connection Management document needs to be updated to NEST 3.0
**Describe the flaw**
The [Connection Management guide](https://nest-simulator.readthedocs.io/en/latest/guides/connection_management.html) on ReadTheDocs is not updated to NEST 3.0. Some of the code-snippets won't work anymore, and there are new examples that should be added. It also seems quite outdated, one of the first lines are
```
In SLI ,the old syntax of the function still works, while in PyNEST, the Connect() function has been renamed to OneToOneConnect().
```
but there is no `OneToOneConnect()` in Python, we just have `Connect()`.

### PR DESCRIPTION
**Pull request by stinebuu**
I have updated the ``connection_management`` documentation to fit with changes in NEST 3.0.
@sarakonradi, feel free to make changes.

This fixes #1675.

### CODE REVIEW

File: doc/guides/connection_management.rst [hunk 1]
```diff
@@ -3,35 +3,24 @@
...
-under the key 'rule'. The patterns available are described in
-`Connection Rules <connection-management.md#connection-rules>`__. In
-addition the synapse model can be specified within the connect function
-and all synaptic parameters can be randomly distributed.
-
-The `Connect()` function can be called in either of the following
-manners:
+In NEST, connections are created with the `Connect()` function. You can
+create connections with different types of connectivity patterns, which
+are defined inside the function under the ``conn_spec`` argument, and the
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:8]:
```suggestion
are defined inside the function under the ``conn_spec`` argument and the
```
```suggestion
are defined inside the function under the ``conn_spec`` argument, and the
```

File: doc/guides/connection_management.rst [hunk 2]
```diff
@@ -3,35 +3,24 @@
...
+key 'rule'. The patterns available are described in
+:ref:`Connection Rules <conn_rules>`. In addition the synapse model can
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:10]:
```suggestion
key 'rule'. The patterns available are described in the
:ref:`Connection Rules <conn_rules>` section. In addition, the synapse model can
```
```suggestion
key 'rule'. The patterns available are described in
:ref:`Connection Rules <conn_rules>`. In addition the synapse model can
```

File: doc/guides/connection_management.rst [hunk 3]
```diff
@@ -3,35 +3,24 @@
...
+be specified within the Connect function and all synaptic parameters can
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:11]:
```suggestion
be specified within the `Connect()` function, and all synaptic parameters can
```

File: doc/guides/connection_management.rst [hunk 4]
```diff
@@ -3,35 +3,24 @@
...
+be randomly distributed. More information about synapse models and their
+parameters can be found in :ref:`Synapse Specification <synapse_spec>`.
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:13]:
```suggestion
parameters can be found in the :ref:`Synapse Specification <synapse_spec>` section.
```

File: doc/guides/connection_management.rst [hunk 5]
```diff
@@ -42,20 +31,18 @@ be given.
 In addition switches allowing self-connections (``allow_autapses``, default:
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:31]:
```suggestion
In addition, switches allowing self-connections (``allow_autapses``, default:
```

File: doc/guides/connection_management.rst [hunk 6]
```diff
@@ -42,20 +31,18 @@ be given.
...
-vector. Here 'allow_multapses' set to False will result in one potential
-connection between each occurring node pair.
+particular connection might still be established multiple times.
 
 ``syn_spec`` defines the synapse type and its properties. It can be
 given as a string defining the synapse model (default:
-'static_synapse') or as a dictionary. By using the key-word variant
-(``Connect(pre, post, syn_spec=syn_spec_dict)``), the conn_spec can be
+'static_synapse'), as an object defining :ref:`collocated synapses <collocated_synapses>`,
+or as a dictionary. By using the key-word variant
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:41]:
```suggestion
or as a dictionary. By using the keyword variant
```

File: doc/guides/connection_management.rst [hunk 7]
```diff
@@ -42,20 +31,18 @@ be given.
...
+(``Connect(pre, post, syn_spec=syn_spec_dict)``), the ``conn_spec`` can be
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:42]:
```suggestion
(``Connect(pre, post, syn_spec=syn_spec_dict)``), ``conn_spec`` can be
```

File: doc/guides/connection_management.rst [hunk 8]
```diff
@@ -69,38 +56,23 @@ given as strings, for all other rules, a dictionary specifying the rule
...
-~~~~~~~~~~~~
+~~~~~~~~~~
 
 .. image:: ../_static/img/One_to_one.png
      :width: 200px
      :align: center
 
 The ith node in ``pre`` is connected to the ith node in ``post``. The
-node lists pre and post have to be of the same length.
+NodeCollections of pre and post have to be of the same length.
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:66]:
```suggestion
NodeCollections of ``pre`` and ``post`` have to be of the same length.
```

File: doc/guides/connection_management.rst [hunk 9]
```diff
@@ -189,38 +161,57 @@ Example:
...
-    CopyModel("static_synapse","excitatory",{"weight":2.5, "delay":0.5})
-    Connect(A, B, syn_spec="excitatory")
+    A = nest.Create('iaf_psc_alpha', n)
+    B = nest.Create('iaf_psc_alpha', n)
+    nest.CopyModel('static_synapse','excitatory',{'weight': 2.5, 'delay': 0.5})
+    nest.Connect(A, B, syn_spec='excitatory')
 
 Specifying the synapse properties in a dictionary allows for distributed
-synaptic parameter. In addition to the key ``model`` the dictionary can
+synaptic parameters. In addition to the key ``synapse_model`` the dictionary can
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:209]:
```suggestion
synaptic parameters. In addition to the key ``synapse_model``, the dictionary can
```

File: doc/guides/connection_management.rst [hunk 10]
```diff
@@ -234,144 +225,196 @@ Example:
...
-    A = Create("iaf_psc_alpha", 2)
-    B = Create("iaf_psc_alpha", 5)
+    A = nest.Create('iaf_psc_alpha', 2)
+    B = nest.Create('iaf_psc_alpha', 5)
     conn_dict = {'rule': 'fixed_outdegree', 'outdegree': 3}
     syn_dict = {'weight': [[1.2, -3.5, 0.4], [-0.2, 0.6, 2.2]]}
-    Connect(A, B, conn_spec=conn_dict, syn_spec=syn_dict)
+    nest.Connect(A, B, conn_spec=conn_dict, syn_spec=syn_dict)
+
+For ``fixed_total_number`` the array has to be same the length as the
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:294]:
```suggestion
For ``fixed_total_number``, the array has to be same the length as the
```

File: doc/guides/connection_management.rst [hunk 11]
```diff
@@ -234,144 +225,196 @@ Example:
...
+
+    syn_spec = nest.CollocatedSynapses({'weight': 5.0, 'receptor_type': 2},
+                                       {'weight': 1.5, 'receptor_type': 7})
+
+    nest.Connect(src, trgt, 'one_to_one', syn_spec=syn_spec)
+
+    conns = nest.GetConnections()
+    print(conns.get())
+
+You can see how many synapse parameters you have by doing `len()` on your `CollocatedSynapses` object:
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:351]:
```suggestion
You can see how many synapse parameters you have by calling `len()` on your `CollocatedSynapses` object:
```

File: doc/guides/connection_management.rst [hunk 12]
```diff
@@ -234,144 +225,196 @@ Example:
...
+The following parameters and functionalities are provided:
+
+- Random parameters
+- Spatial parameters
+- Spatially distributed parameters
+- Mathematical functions
+- Clipping, redrawing, and conditional parameters
+
+For more information, check out the documentation on the different
+:doc:`PyNEST api's <../ref_material/pynest_apis>`, or the section on parametrization
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:377]:
```suggestion
:doc:`PyNEST APIs <../ref_material/pynest_apis>`, or the section on parametrization
```

File: doc/guides/connection_management.rst [hunk 13]
```diff
@@ -234,144 +225,196 @@ Example:
...
+in our :doc:`nest2_to_nest3/nest2_to_nest3_overview` guide.
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:378]: Should we add a ``:ref:`` to the section?

File: doc/guides/connection_management.rst [hunk 14]
```diff
@@ -393,69 +436,134 @@ post-synaptic node. Receptor types are identified by integer numbers,
...
-connection setup. The default synapse type in NEST is
+NEST supports multiple synapse types that can be specified during
+connection setup. The default synapse type in NEST is the
 ``static_synapse``. Its weight does not change over time. To allow
 learning and plasticity, it is possible to use other synapse types that
 implement long-term or short-term plasticity. A list of available types
-is accessible via the command ``Models("synapses")``. The output of this
-command (as of revision 11199) is shown below:
+is accessible via the command ``nest.Models('synapses')``. The output of this
+command (as of commit ``1159664``) is shown below:
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:488]: Should we link the commit?

File: doc/guides/connection_management.rst [hunk 15]
```diff
@@ -464,34 +572,36 @@ synapses of a type. This means that these are the same for all
...
 the defaults of the copied model are taken.
 
 ::
 
-    CopyModel("static_synapse", "inhibitory", {"weight": -2.5})
-    Connect(A, B, syn_spec="inhibitory")
+    nest.CopyModel('static_synapse', 'inhibitory', {'weight': -2.5})
+    nest.Connect(A, B, syn_spec='inhibitory')
 
 **Note**: Not all nodes can be connected via all available synapse
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:606]: We could turn this into an RST note.

File: doc/guides/connection_management.rst [hunk 16]
```diff
@@ -189,38 +161,57 @@ Example:
...
 
 Specifying the synapse properties in a dictionary allows for distributed
-synaptic parameter. In addition to the key ``model`` the dictionary can
+synaptic parameters. In addition to the key ``synapse_model`` the dictionary can
 contain specifications for ``weight``, ``delay``, ``receptor_type`` and
 parameters specific to the chosen synapse model. The specification of
 all parameters is optional. Unspecified parameters will use the default
 values determined by the current synapse model. All parameters can be
-scalars, arrays or distributions (specified as dictionaries). One
+scalars, arrays or distributions (specified as a ``nest.Parameter``). One
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:214]: Perhaps we could have parameter subsections titled "scalar", "array" and "distributed". What do you think?

**Review by sarakonradi — CHANGES_REQUESTED** [review]
Thanks a lot for updating the documentation! Please find my comments inline.

File: doc/guides/connection_management.rst (hunk 13 above)
Comment (stinebuu) [reply doc/guides/connection_management.rst:378]: Yes! I wanted to do that, I just didn't know how, I will update :)

File: doc/guides/connection_management.rst (hunk 16 above)
Comment (stinebuu) [reply doc/guides/connection_management.rst:214]: Will do!

**Comment by stinebuu** [comment]
Thanks @sarakonradi! I have now applied your suggestions. I have also updated the sections on inspecting and changing the connections to include `SynapseCollection` functionalities. It was only documented in the nest 2 -> nest 3 documentation, but it should be documented in the connection_management guide as well.

File: doc/guides/connection_management.rst [hunk 17]
```diff
@@ -44,6 +44,16 @@ omitted in the call to connect and 'all_to_all' is assumed as the
 default. The exact usage of the synapse dictionary is described in
 :ref:`synapse_spec`.
 
+After a connection is established it might be beneficial to look up the number of
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:47]:
``` suggestion
After a connection is established, it might be beneficial to look up the number of
```

File: doc/guides/connection_management.rst [hunk 18]
```diff
@@ -629,6 +642,65 @@ parameters of the connections:
          'target_thread': [0, 0],
          'weight': [1.0, 1.0]}
 
+The ``get()`` function also takes a string or list of strings as argument. You
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:645]:
``` suggestion
The ``get()`` function also takes a string or list of strings as arguments. You
```

File: doc/guides/connection_management.rst [hunk 19]
```diff
@@ -629,6 +642,65 @@ parameters of the connections:
...
+      0.4714045207910317,
+      0.33333333333333337,
+      3.925231146709438e-17,
+      0.33333333333333326,
+      0.4714045207910317,
+      0.33333333333333326,
+      0.47140452079103157)
+
+You can further examine the SynapseCollection by checking the length of the object.
+Or, by printing it, which will return a table of source and target node IDs:
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:676]:
``` suggestion
You can further examine the SynapseCollection by checking the length of the object
or by printing it, which will return a table of source and target node IDs:
```

File: doc/guides/connection_management.rst [hunk 20]
```diff
@@ -659,3 +732,29 @@ by using the ``set()`` function on the SynapseCollection:
          'target': 2,
          'target_thread': 0,
          'weight': 2.0}
+
+Updating a single parameter is done by calling ``set(parameter_name=parameter_value)``.
+You can use a single value, a list, or a ``nest.Parameter`` as value. If a single
+value is given, the value is set on all connections, while if you use a list to set
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:738]:
``` suggestion
value is given, the value is set on all connections. If you use a list to set
```

File: doc/guides/connection_management.rst [hunk 21]
```diff
@@ -659,3 +732,29 @@ by using the ``set()`` function on the SynapseCollection:
...
+the parameter, the list needs to be the same length as the SynapseCollection.
+
+  >>>  conn.set(weight=[4.0, 4.5, 5.0, 5.5])
+
+Just as you can retrieve several parameters at once with the ``get()`` function
+above, you can also set several parameters at once, with
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:744]:
``` suggestion
above, you can also set several parameters at once with
```

File: doc/guides/connection_management.rst [hunk 22]
```diff
@@ -659,3 +732,29 @@ by using the ``set()`` function on the SynapseCollection:
...
+``set(parameter_dictionary)``. You can again use a single value, a list, or a
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:745]:
``` suggestion
``set(parameter_dictionary)``. Again, you can again use a single value, a list, or a
```

File: doc/guides/connection_management.rst [hunk 23]
```diff
@@ -659,3 +732,29 @@ by using the ``set()`` function on the SynapseCollection:
...
+You can also directly set parameters of your SynapseCollection:
+
+  >>>  conn.weight = 5.
+  >>>  conn.weight
+       [5.0, 5.0, 5.0, 5.0]
+  >>>  conn.delay = [5.1, 5.2, 5.3, 5.4]
+  >>>  conn.delay
+       [5.1, 5.2, 5.3, 5.4]
+
+Note that some parameters, like `source` and `target`, cannot be set.  The documentation of a specific
```
Comment (sarakonradi) [inline doc/guides/connection_management.rst:759]:
``` suggestion
Note that some parameters, like `source` and `target`, cannot be set. The documentation of a specific
```

**Review by sarakonradi — CHANGES_REQUESTED** [review]
> Thanks @sarakonradi! I have now applied your suggestions. I have also updated the sections on inspecting and changing the connections to include SynapseCollection functionalities. It was only documented in the nest 2 -> nest 3 documentation, but it should be documented in the connection_management guide as well.

Thanks for that! Good idea to document it in this guide as well. 

I have added a few more minor comments.

**Comment by stinebuu** [comment]
@sarakonradi @jougs I have added an index on the right side, and updated the headers a bit so they are consistent throughout the document :)

**Review by jougs — CHANGES_REQUESTED** [review]
Here's a first set of suggestions. However, as I find it quite hard to follow the flow of the file with all the displayed changes, I'll file another set in the form of a PR against your branch once these are addressed, if that's fine by you.

File: doc/guides/connection_management.rst [hunk 24]
```diff
@@ -69,38 +70,21 @@ given as strings, for all other rules, a dictionary specifying the rule
...
-::
-
-    weight = 1.5
-    delay = 0.5
-    Connect(A[0], B[0], weight, delay)
+    A = nest.Create('iaf_psc_alpha', n)
+    B = nest.Create('spike_recorder', n)
+    nest.Connect(A, B, 'one_to_one')
 
 all-to-all
```
[the remark this answers is not in the document]
Comment (stinebuu) [reply doc/guides/connection_management.rst:85]: Done in [1b80b22](https://github.com/nest/nest-simulator/pull/1834/commits/1b80b22bc0996a9570cbea9c58a76e2794582ff2).

File: doc/guides/connection_management.rst [hunk 25]
```diff
@@ -152,230 +132,284 @@ fixed-outdegree
...
 
-Example:
+Array parameters
+~~~~~~~~~~~~~~~~
+
+Array parameters can be used in conjunction with the rules
+``one_to_one``, ``all_to_all``, ``fixed_indegree``, ``fixed_outdegree``
+and ``fixed_total_number``. The arrays can be specified as numpy arrays or
+lists. As with the scalar parameters, all parameters but the receptor
+types must be specified as arrays of floats. For ``one_to_one`` the
```
[the remark this answers is not in the document]
Comment (stinebuu) [reply doc/guides/connection_management.rst]: Done in [1b80b22](https://github.com/nest/nest-simulator/pull/1834/commits/1b80b22bc0996a9570cbea9c58a76e2794582ff2).

File: doc/guides/connection_management.rst [hunk 26]
```diff
@@ -152,230 +132,284 @@ fixed-outdegree
...
+    nest.CopyModel('static_synapse','excitatory',{'weight': 2.5, 'delay': 0.5})
+    nest.Connect(A, B, syn_spec='excitatory')
 
 Specifying the synapse properties in a dictionary allows for distributed
-synaptic parameter. In addition to the key ``model`` the dictionary can
+synaptic parameters. In addition to the key ``synapse_model``, the dictionary can
 contain specifications for ``weight``, ``delay``, ``receptor_type`` and
 parameters specific to the chosen synapse model. The specification of
 all parameters is optional. Unspecified parameters will use the default
 values determined by the current synapse model. All parameters can be
```
[the remark this answers is not in the document]
Comment (stinebuu) [reply doc/guides/connection_management.rst]: Done in [0239897](https://github.com/nest/nest-simulator/pull/1834/commits/02398970ea954faf592d2c23952cf7c61048f505).

**Comment by stinebuu** [comment]
@jougs Thanks for your comments, and sorry for taking some time to address them. Feel free to create a PR towards the branch with all your changes ;)

**Comment by jougs** [comment]
@stinebuu: No worries!

> Feel free to create a PR towards the branch with all your changes ;)

I am on it, but afraid that I can only send you something by the end of the week. This guide is just crazy long and some of the parts are really horrible (especially the ones I myself wrote 14 years ago...). Anyway, stay tuned!

**Review by jougs — APPROVED** [review]
Awesome!

I've just restarted the Travis build for MacOS, which failed due to a timeout, but propose to merge anyway if it fails again, as this does not actually touch any code.

**Comment by jougs** [comment]
@sarakonradi: could you now please give this a read? Thanks!

**Comment by stinebuu** [comment]
I think the Travis failure will be fixed with #1889.

**Comment by jougs** [comment]
I've just merged #1889. Please pull master to see if that has the desired effect. Thanks!

**Comment by stinebuu** [comment]
Travis pass!

**Review by sarakonradi — APPROVED** [review]
Amazing. Thank you both for the many updates!

The only aspects I noticed (e.g., CSS) have are already discussed in stinebuu/nest-simulator#8 and to be addressed separately. The rest looks good to me.

**Comment by jougs** [comment]
Thanks to all of you! Merging!
